src/rgw/rgw_op.cc

```diff
@@ -1145,9 +1150,500 @@ int RGWGetObj::get_data_cb(bufferlist& bl, off_t bl_ofs, off_t bl_len)
+
+  char *pstr = bl.c_str();
+  char sha[25];
+
+  /* get sha1 */
+  for (off_t i = 0; i < num; i++)
+  {
+    memset(sha, 0x00, sizeof(sha));
+    h->Update((byte *)pstr, info.piece_length);
+    h->Final((byte *)sha);
+
+    set_info_pieces(sha);
+    pstr += info.piece_length;
+  }
+
+  /* process remain */
+  if (0 != remain)
+  {
+    memset(sha, 0x00, sizeof(sha));
+
+    h->Update((byte *)pstr, remain);
+    h->Final((byte *)sha);
+
+    set_info_pieces(sha);
+  }
+}
+
+int seed::sha1_process()
+{
+  uint64_t remain = info.len%info.piece_length;
+  uint8_t  remain_len = ((remain > 0)? 1 : 0);
+  sha_len = (info.len/info.piece_length + remain_len)*CEPH_CRYPTO_SHA1_DIGESTSIZE;
+
+  SHA1 h;
+  list<bufferlist>::iterator iter = torrent_bl.begin();
+  for (; iter != torrent_bl.end(); iter++)
+  {
+    bufferlist &bl_info = *iter;
+    sha1(&h, bl_info, (*iter).length());
+  }
+
+  return 0;
+}
+
+int seed::handle_data()
+{
+  int ret = 0;
+
+  /* sha1 process */
+  ret = sha1_process();
+  if (0 != ret)
+  {
+    INFO "sha1_process fail ret= "<<ret<<dendl;
+    return ret;
+  }
+
+  /* produce torrent data */
+  do_encode();
+
+  /* save torrent data into OMAP */
+  ret = save_torrent_file();
+  if (0 != ret)
+  {
+    INFO "save_torrent fail ret= "<<ret<<dendl;
+    return ret;
+  }
+
+  return 0;
+}
+
+int seed::get_params()
+{
+  if ((g_conf->torrent_tracker).empty() && (g_conf->torrent_origin).empty())
+  {
+    INFO "torrent_tracker and torrent_origin can not be empty at the same time in ceph.conf" << dendl;
+    return -1;
+  }
+
+  is_torrent = true;
+  info.piece_length = g_conf->sha_unit;
+  create_by = g_conf->torrent_createby;
+  set_announce(g_conf->torrent_tracker);
+  encoding = g_conf->torrent_encoding;
+  origin = g_conf->torrent_origin;
+  comment = g_conf->torrent_comment;
+
+  return 0;
+}
+
+void seed::set_announce(string value)
+{
+  if (value.empty())
+  {
+    INFO "torrent_tracker is empty" << dendl;
+    return;
+  }
+  const char *pstr = value.c_str();
+  char *pstart = const_cast<char *>(pstr);
+  uint64_t len = value.length();
+  char *pend = pstart + len;
+  char *location;
+  char flag = ',';
+
+  list<string> announce_list;  // used to get announce list from conf
+  while (pstart < pend)
+  {
+    location = NULL;
+    location = strchr(pstart, flag);
+    if (!location)
+    {
+      string ann(pstart);
+      announce_list.push_back(ann);
+      break;
+    }
+
+    char temp[100] = { 0 };
+    snprintf(temp, location - pstart + 1, "%s", pstart);
+    string anno(temp);
+    announce_list.push_back(anno);
+    pstart = location + 1;
+  }
+
+  list<string>::iterator iter = announce_list.begin();
+  announce = *iter;
+  iter++;
+
+  for (; iter != announce_list.end(); ++iter)
+  {
+    listvalue info_list;
+    info_list.type = BE_LIST;
+    info_list.info =  (*iter);
+    info_list.field = "";
+    list_info.push_back(info_list);
+  }
+  announce_list.clear();
+}
+
+void seed::do_encode()
```

Comment:
i would prefer to see `do_encode()` encode each field directly into the bufferlist, rather than building up a std::list and processing it all at once with `seed_encode()`. there are a lot of allocations associated with std::list that we should be able to avoid here